File: apps/api/src/application/import/allure.parser.ts
import type { AllureResultItem } from "@testcraft/types";
import { TestResultStatus } from "@testcraft/types";

import type { ParsedTestCase } from "./import.repository";

const resolveStatus = (
  status: AllureResultItem["status"],
): TestResultStatus => {
  switch (status) {
    case "passed": {
      return TestResultStatus.Passed;
    }
    case "failed":
    case "broken": {
      return TestResultStatus.Failed;
    }
    case "skipped": {
      return TestResultStatus.Skipped;
    }
    default: {
      return TestResultStatus.Blocked;
    }
  }
};

const labelValue = (
  labels: AllureResultItem["labels"],
  ...keys: string[]
): string | null => {
  if (!labels) return null;

  const map = new Map(labels.map((l) => [l.name, l.value]));

  for (const key of keys) {
    const val = map.get(key);

    if (val) return val;
  }

  return null;
};

export const parseAllure = (results: AllureResultItem[]): ParsedTestCase[] =>
  results.map((result, index) => ({
    suiteName:
      labelValue(result.labels, "suite", "parentSuite", "testClass") ??
      "Default Suite",
    caseName: result.name ?? result.fullName ?? `Unknown (${index + 1})`,
    status: resolveStatus(result.status),
    notes: result.statusDetails?.message ?? null,
  }));

File: apps/Api/src/TestCraft.Application/Import/AllureParser.cs
using TestCraft.Domain.Enums;

namespace TestCraft.Application.Import;

public static class AllureParser
{
    public static List<ParsedTestCase> Parse(IReadOnlyList<AllureResultItem> results)
    {
        var cases = new List<ParsedTestCase>(results.Count);

        for (var i = 0; i < results.Count; i++)
        {
            var result = results[i];
            cases.Add(
                new ParsedTestCase
                {
                    SuiteName =
                        LabelValue(result.Labels, "suite", "parentSuite", "testClass")
                        ?? "Default Suite",
                    CaseName = result.Name ?? result.FullName ?? $"Unknown ({i + 1})",
                    Status = ResolveStatus(result.Status),
                    Notes = result.StatusDetails?.Message,
                }
            );
        }

        return cases;
    }

    private static TestResultStatus ResolveStatus(string? status) =>
        status switch
        {
            "passed" => TestResultStatus.Passed,
            "failed" or "broken" => TestResultStatus.Failed,
            "skipped" => TestResultStatus.Skipped,
            _ => TestResultStatus.Blocked,
        };

    private static string? LabelValue(IReadOnlyList<AllureLabel>? labels, params string[] keys)
    {
        if (labels is null)
        {
            return null;
        }

        var map = new Dictionary<string, string>();
        foreach (var label in labels)
        {
            map[label.Name] = label.Value;
        }

        foreach (var key in keys)
        {
            if (map.TryGetValue(key, out var value) && !string.IsNullOrEmpty(value))
            {
                return value;
            }
        }

        return null;
    }
}
